### DISCUSSION TITLE
!!! BREAKING CHANGES - PhET APIs !!!

### DISCUSSION THREAD

**Discussion by brettfiedler** [phetsims/community#22] [Announcements]
PhET APIs are constantly evolving and improving. 

In our current structure, that sometimes means making changes that are not backward compatible. We know that third-party developers are using our libraries, and breaking changes are a drag - especially if you don’t know about them!

Watch this thread where we will post notifications of breaking changes to our APIs. You can subscribe to this thread in the sidebar under the Notifications header.

**Comment by brettfiedler** [comment]
### Breaking changes were formerly announced in a Google Group thread, [archived here](https://groups.google.com/g/developing-interactive-simulations-in-html5/c/1Zm4l3sye20).

2023 posts for Breaking Changes:

#### Jan 2, 2023 - AXON
> In https://github.com/phetsims/axon/issues/425, NumberProperty's `range` option was changed to no longer accept `null` as a value. Now all NumberProperty instances have a defined range, defaulting to (-∞,∞). Thus NumberProperty.prototype.range and NumberProperty.prototype.rangeProperty.value will always be of type Range.

#### Feb 1, 2023 - AXON
> In https://github.com/phetsims/axon/issues/428, we removed the `ReadOnlyProperty.useDeepEquality` option because it was confusing and limited in scope. Instead, use `valueComparisonStrategy: 'eqaualsFunction' to reproduce the same functionality as `useDeepEquality`. 

#### Feb 3, 2023 - SUN
> In https://github.com/phetsims/sun/issues/814, we have changed the public API for the sun Carousel. The changes help Carousel have better dynamic support, for adding and removing items.  The API change is that instead of accepting an array of Nodes, it accepts an array of CarouselItem.

#### Mar 1, 2023 - SCENERY
> In https://github.com/phetsims/scenery/issues/1472, we renamed the "text" related API to "string" when talking about the string that the node renders. This is for Text and RichText:
> 
> * The Node option `text` is now `string`
> * getText() and setText() are now getString() and setString()
> * es5 getters and setters for text are now string
> 
> Please report in that issue if you run into trouble.

**Comment by brettfiedler** [comment]
**April 13, 2023 - CHIPPER**
> PhET is now using TypeScript 5.0, please see https://github.com/phetsims/chipper/issues/1384 for details.  Please pull chipper and npm install to get the new version.  After updating, you can run `cd chipper` then `node node_modules/typescript/bin/tsc --version` to see that your version updated correctly. Please report problems in the issue.

**Comment by zepumph** [comment]
**Aug 16, 2023 - EVERYWHERE**

We have recently transferred our primary branch from `master` over to `main`. Here are some notes for developers to change their working copies:

> The transition from "master" to "main" is, complete. The team that worked on this thought it would be helpful to include a set of instructions for what to do it one finds themself on a master branch, or with all master branches checked out.
> 
> Here is what to do if you find yourself locally on the master branch for everything.
> 
> 1. Make sure you have a clean, freshly pulled, working copy. If not stash any local changes and make sure you're pulled.
> 2. Run `perennial/bin/for-each.sh active-repos git checkout main; perennial/bin/for-each.sh active-repos git symbolic-ref refs/remotes/origin/HEAD refs/remotes/origin/main` to switch all active repos to the main branch.  If you have other repos that are not in the list of active ones you can run this command instead (your mileage may vary): `for f in $(ls) ; do (cd $f; git pull ; git checkout main; git symbolic-ref refs/remotes/origin/HEAD refs/remotes/origin/main) ; done`
> 3. Run `perennial/bin/for-each.sh perennial/data/active-repos "rm .git/hooks/pre-commit; git init --template=../phet-info/git-template-dir"` to update your git hooks to run on the "main" branch.
> 4. Double check that all is good. Run `perennial/bin/for-each.sh active-repos git status` and make sure each says you're pointing to `origin/main` and not `origin/master`.
> 
> That's it. You should be good to go.

**Comment by brettfiedler** [comment]
See https://github.com/orgs/scenerystack/discussions/9 for the latest BREAKING CHANGES in the new SceneryStack community!
